ruby/phase-1.md

```diff
@@ -0,0 +1,398 @@
+can choose to gradually start overriding the defaults until they achieve what feels comfortable.*
+
+
+## Client
+
+The client is the primary point for interacting with workflows and async activities as well as
+managing namespaces. As hinted in the previous section there's a global singleton client available
+for applications that only connect to a single Temporal server and don't need varying
+configurations:
+
+```ruby
+# Global instance of the Client configured via `Temporal.configure`
+Temporal.start_workflow(MyWorkflow)
+
+# Explicitly configured local instance of the Client
+config = Temporal::Configuration.new(url: 'localhost:7233')
+client = Temporal::Client.new(config)
+client.start_workflow(MyWorkflow)
+```
+
+Despite the differences both clients offer exactly the same interface (the singleton client is an
+instance of a regular client stored on Temporal module for convenience). The interface will look
+something like this:
+
+```ruby
+class Client
+  def initialize(config)
+
+  def start_workflow(
+    workflow, # workflow class or a name
+    *args, # any args to be passed as the input to the workflow
+    options: {}, # execution options overrides
+    **kwargs # keyword arguments as inputs are also supported
+  ) -> WorkflowHandle
+
+  def workflow_handle(namespace, workflow_id, run_id = nil) -> WorkflowHandle
+end
+```
+
+Notes:
+
+- This is not a complete definition of the client
+- Both named and keyword arguments are supported, except for the reserved `options:`
+
+The combination of the singleton client and implicit execution attributes allows us to optionally
+support direct workflow invocations (removing all SDK traces out of the way):
+
+```ruby
+MyWorkflow.start(arg_1, arg_2, keyword_arg: arg_3)
+```
+
+Once a workflow is started all the interactions will be going through the handle object returned to
+the caller. The handle can of course be created from a namespace/workflow_id/run_id combination
+using the `Client#workflow_handle` method. Here the interface of the handle:
+
+```ruby
+class WorkflowHandle
+  def initialize(connection, namespace, workflow_id, run_id = nil)
+
+  def result(await: false, timeout: nil) -> Any
+
+  def describe -> WorkflowExecutionInfo
+
+  def cancel -> void
+
+  def query(name, *args) -> Any
+
+  def signal(name, *args) -> void
+
+  def terminate(reason, details: nil) -> void
+end
+
+```
+
+
+## Worker
+
+The worker interface allows SDK users to register any number of both workflows and activities. These
+will then be polled for on the specified namespaces & task queues and executed. The worker can poll
+on multiple task queues at the same time. These will be determined by the registered activities and
+workflow. If there are any overlaps we'll only create one poller for each unique task queue.
+
+```ruby
+config Temporal::Configuration.new(url: 'localhost:7233')
+worker = Temporal::Worker.new(config)
+
+worker.register_workflow(NewWorkflow)
```

Comment:
We need a `#register_workflows` method too, e.g.
```
worker.register_workflows([Workflow1, Workflow2, Workflow3], 
  task_queue: 'tq-A', namespace: 'foo')
```
Probably okay to omit the array brackets and use `*` to collect the args if added individually. But would also be nice to be able to just pass in a named array (pulled from some application config elsewhere) to register all of them in one call.

Reply:
Ah, yes, that's a good optimisation. Another approach I had in mind is auto-registration. We had a few incidents at CB when someone would implement a workflow/activity, but forget to register it with the worker. However I'm not yet sure how that would work